### PR TITLE
Always use toAbsolutePath() for MountableFile

### LINKED ISSUE #3493: Parsing error after upgrading docker version on Windows
Hi,

I am working on java testcontainers (version 1.14.3) on Windows. recently, my docker desktop version was upgraded from 2.4.0.0 to 2.5.0.0. since then I am getting those exceptions on initialization:

```
Caused by: org.testcontainers.shaded.com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `com.github.dockerjava.api.model.Binds`, **problem: Error parsing Bind 'C:/dev/IM2/BackendTests/Tests/../run/community-eval/logs/:/service/community-eval/logs:rw'**
 at [Source: (org.testcontainers.shaded.okio.RealBufferedSource$1); line: 1, column: 1272] (through reference chain: com.github.dockerjava.api.command.InspectContainerResponse["HostConfig"]->com.github.dockerjava.api.model.HostConfig["Binds"])
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.exc.ValueInstantiationException.from(ValueInstantiationException.java:47) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.DeserializationContext.instantiationException(DeserializationContext.java:1735) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.wrapAsJsonMappingException(StdValueInstantiator.java:491) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.rewrapCtorProblem(StdValueInstantiator.java:514) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator._createUsingDelegate(StdValueInstantiator.java:551) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createUsingDelegate(StdValueInstantiator.java:298) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createUsingArrayDelegate(StdValueInstantiator.java:307) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromArray(BeanDeserializerBase.java:1435) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:185) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:161) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:138) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:138) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4218) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3259) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.okhttp.OkHttpInvocationBuilder.get(OkHttpInvocationBuilder.java:90) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.InspectContainerCmdExec.execute(InspectContainerCmdExec.java:31) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.InspectContainerCmdExec.execute(InspectContainerCmdExec.java:13) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.exec.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:21) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:35) ~[testcontainers-1.14.3.jar:?]
	at com.github.dockerjava.core.command.InspectContainerCmdImpl.exec(InspectContainerCmdImpl.java:52) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:428) ~[testcontainers-1.14.3.jar:?]
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:325) ~[testcontainers-1.14.3.jar:?]
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81) ~[duct-tape-1.0.8.jar:?]
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:323) ~[testcontainers-1.14.3.jar:?]
	... 24 more
**Caused by: java.lang.IllegalArgumentException: Error parsing Bind 'C:/dev/IM2/BackendTests/Tests/../run/community-eval/logs/:/service/community-eval/logs:rw'
	at com.github.dockerjava.api.model.Bind.parse(Bind.java:134) ~[testcontainers-1.14.3.jar:?]**
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:1.8.0_271]
	at java.util.Spliterators$ArraySpliterator.forEachRemaining(Unknown Source) ~[?:1.8.0_271]
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:1.8.0_271]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:1.8.0_271]
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:1.8.0_271]
	at java.util.stream.AbstractPipeline.evaluateToArrayNode(Unknown Source) ~[?:1.8.0_271]
	at java.util.stream.ReferencePipeline.toArray(Unknown Source) ~[?:1.8.0_271]
	at com.github.dockerjava.api.model.Binds.fromPrimitive(Binds.java:30) ~[testcontainers-1.14.3.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_271]
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_271]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_271]
	at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_271]
```

It seems that Bind.parse method does not support Windows path at all. 
However, in the previous docker desktop version(2.4.0.0) the Windows path was translated to /host_mnt/C/... which can be parsed by that function. 

any idea how can this be resolved in new docker desktop version?

### PR DESCRIPTION
**Pull request by kiview**
`forHostPath` in `MountableFile` will now always use `toAbsolutePath()`, instead of using `toURI()` in certain cases.
Solves an issue on Windows that was introduced with the latest Docker Desktop on Windows update.

Fixes #3493.

### CODE REVIEW

**Comment by Fufusulier** [comment]
Hello,
is there a reason why this PR is still opened?
Thanks

**Comment by bsideup** [comment]
@Fufusulier well, "Testcontainers isn't our primary job" is probably the best answer to "a reason why this PR is still opened" 😅 

We're on it, just it takes time to handle everything. Stay tuned 👍

**Comment by Fufusulier** [comment]
> @Fufusulier well, "Testcontainers isn't our primary job" is probably the best answer to "a reason why this PR is still opened" 😅
> 
> We're on it, just it takes time to handle everything. Stay tuned 👍

oooh ok, I thought it was. sorry :)

**Comment by bsideup** [comment]
@Fufusulier no problem! We always take it as a compliment :)

We thought that the issue is in docker-java, but apparently it wasn't, so it takes a bit longer to fix it, but at least we identified the root cause and working on it 👍

File: core/src/main/java/org/testcontainers/containers/GenericContainer.java [hunk 1]
```diff
@@ -946,7 +946,7 @@ public void addEnv(String key, String value) {
     public void addFileSystemBind(final String hostPath, final String containerPath, final BindMode mode, final SelinuxContext selinuxContext) {
 
         final MountableFile mountableFile = MountableFile.forHostPath(hostPath);
-        binds.add(new Bind(mountableFile.getResolvedPath(), new Volume(containerPath), mode.accessMode, selinuxContext.selContext));
+        binds.add(new Bind(mountableFile.getUnixFilesystemPath(), new Volume(containerPath), mode.accessMode, selinuxContext.selContext));
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/containers/GenericContainer.java:949]: I am not too sure we should do it here. Future versions of Docker for Windows may become strict about the path (`C:\foo\bar` is good, `/mnt/c/foo/bar` is bad)

I would just ensure that we pass it as `C:\` and not `C:/`

Comment (kiview) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java:949]: It translates into `/c/foo/bar` btw.

Comment (bsideup) [reply core/src/main/java/org/testcontainers/containers/GenericContainer.java:949]: Ooops, sorry, my memory failed me :) But nevertheless, I would let Docker make the decision on how to translate `C:\foo\bar` into a valid bind

**Review by bsideup — APPROVED** [review]

**Comment by Fufusulier** [comment]
this feels good :p

**Comment by jetersen** [comment]
This change sadly stopped accepting Paths from a class resource.
MountableFile does not accept a class to properly load resources from a reasonable class loader.

```java
.withCopyFileToContainer(
    forHostPath(TestConstants.class.getResource("vaultTest_agent.hcl").getPath()),
    "/agent.hcl")
```

I needed to copy this method since it is marked private 😓
https://github.com/testcontainers/testcontainers-java/blob/61dcd7ce371f84d28fcc28192223d41c20eb95bc/core/src/main/java/org/testcontainers/utility/MountableFile.java#L169-L179

See https://github.com/jenkinsci/hashicorp-vault-plugin/pull/168

**Comment by bsideup** [comment]
@jetersen any reason you're not using `forClasspathResource`?

**Comment by jetersen** [comment]
`forClasspathResource` returns Resources are not found 

As i said it using a wrong class loader.

**Comment by bsideup** [comment]
@jetersen are you sure the problem is with the classloaders? You seem to be providing path `vaultTest_adminPolicy.hcl`, while most probably the actual path is something like `com/example/tests/constants/vaultTest_adminPolicy.hcl` (judging by the `TestConstants` class name). When you do `TestConstants.class.getResource("vaultTest_agent.hcl")`, it will append `TestConstants`' package to the resource path.

**Comment by jetersen** [comment]
Right, that was not obvious for the method name and javadoc that I had to provide the full resource path as the variable is named "resourceName" and not "resourcePath"

**Comment by bsideup** [comment]
I agree that the naming could be imported, ut it is similar to other classpath APIs (eg Spring's ClasspathResource) - you need to provide the full path, since there is no "current directory" when it comes to classpath :)

**Pull request merged**